# AgalychnisUCE
Agalychnis (tree frog) UCE analyses

# Data
UCE data were generated by RAPiD Genomics. I obtained a copy of the data, which are stored here:

`/uufs/chpc.utah.edu/common/home/gompert-group4/data/UCE_data`

This includes UCE data for 62 individiuals (a few were miss-labeled, see notes below).

The data were provided by John Wiens, are available from [OSF](https://osf.io/fzw3x/) and were originally published in [Portik et al. 2023](https://academic.oup.com/mbe/article/40/5/msad109/7151539). I specifically used the data from `Final-Trimmed-Alignments.zip`. This includes 3784 UCE alignments. I extracted the sequence for *Agalychnis callidryas* for each UCE alignment (if present), removed any gaps in the alignment and then used the sequence as our reference. Our reference thus comprises 2028 *Agalychnis callidryas* reference sequences (contigs) (1,052,683 bps total).

I used archived UCE data to create a reference UCE file for alignment. This is in `/uufs/chpc.utah.edu/common/home/gompert-group4/data/UCE_data/Reference`.

The reference was indexed with `samtools` (version 1.16) and `bwa` (version 0.7.17-r1198-dirty). 

```{bash}
cd /uufs/chpc.utah.edu/common/home/gompert-group4/data/UCE_data/Reference
samtools faidx CleanUCE.fasta
bwa index CleanUCE.fasta
```
# Alignment and variant calling

The data were demultiplexed, filtered for quality and trimmed by RAPiD Genomics. I am working with the clean files from `/uufs/chpc.utah.edu/common/home/gompert-group4/data/UCE_data/PhylucePipeline/clean-fastq`. 

I aligned the data to the reference UCE set described above with mem algorithm from `bwa` (version 0.7.17-r1198-dirty).

```{bash}
#!/bin/sh
#SBATCH --time=96:00:00
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --account=wolf-kp
#SBATCH --partition=wolf-kp
#SBATCH --job-name=bwa
#SBATCH --mail-type=FAIL
#SBATCH --mail-user=zach.gompert@usu.edu

module load bwa 

cd /uufs/chpc.utah.edu/common/home/gompert-group4/data/UCE_data/Alignments

perl BwaAlignFork.pl CSN_1583*READ1.fastq.gz
```

```{perl}
#!/usr/bin/perl
#
# run bwa mem in parallel
#


use Parallel::ForkManager;
my $max = 16;
my $pm = Parallel::ForkManager->new($max);

my $genome = "/uufs/chpc.utah.edu/common/home/gompert-group4/data/UCE_data/Reference/CleanUCE.fasta";

## get ids
open(IN,"ids.txt") or die "failed to read the ID file\n";
<IN>; # burn header
while(<IN>){
	chomp;
	@line = split(/\s+/,$_);
	$itab{$line[0]} = $line[1];
}
close(IN);

foreach $file1 (@ARGV){
    $pm->start and next; ## fork
    $file2 = $file1;
    $file2 =~ s/READ1/READ2/ or die "failed file sub\n";
    $file1 =~ m/^(CSN_\d+_P\d+_[A-Z0-9]+)/ or die "failed id match\n";
    $pid = $1;
    $id = $itab{$pid};

    print "Working on $pid = $id\n";    
    	
    system "bwa mem -t 1 -M -r 1.3 -k 19 -R \'\@RG\\tID:$id\\tPL:ILLUMINA\\tLB:$id\\tSM:$id\' -o $id".".sam $genome $file1 $file2\n";
    $pm->finish;
}
 

$pm->wait_all_children;
```
I then used `samtools` (version 1.16) to compress, sort and index the sam alignment files, see [Sam2BamFork.pl](Sam2BamFork.pl).

I used `bcftools` (version 1.16) for variant calling as follows:

```{bash}
#!/bin/sh
#SBATCH --time=120:00:00
#SBATCH --nodes=1
#SBATCH --ntasks=24
#SBATCH --account=gompert-kp
#SBATCH --partition=gompert-kp
#SBATCH --job-name=bcf_call
#SBATCH --mail-type=FAIL
#SBATCH --mail-user=zach.gompert@usu.edu


module load samtools
## version 1.16
module load bcftools
## version 1.16

cd /uufs/chpc.utah.edu/common/home/gompert-group4/data/UCE_data/Alignments

bcftools mpileup -b bams -d 1000 -f /uufs/chpc.utah.edu/common/home/gompert-group4/data/UCE_data/Reference/CleanUCE.fasta -a FORMAT/DP,FORMAT/AD -q 20 -Q 30 -I -Ou | bcftools call -v -c -p 0.01 -Ov -o frogs_uce.vcf
```
Variant filtering was performed with [vcfFilter.pl](vcfFilter.pl), which applied a 2X minimim depth and max missing data of 80% along with other quality filters. This left us with 49,028 SNPs in `filtered2x_frogs_uce.vcf`.

We uncovered several labelling errors that were then corrected for downstream analysis. ACA0044 and ACA0039 were dropped; ACA LS 0313, ACA LS0359 and ACA LS 0356 were corrected to ASA, and then ASA LS 0313 as it was already included.

# Population genetic analyses

Everything for the population genetic analyses (and phylogenetic analyses described in the next section) can be found in `/uufs/chpc.utah.edu/common/home/gompert-group4/projects/frogs_spec_contin`.

Working in the `Variants` subdirectory, I converted the filtered vcf file to to gl format and split the gl file by population (this drops the individuals noted above) with [vcf2gl.pl](vcf2gl.pl) and [splitPops.pl](splitPops.pl).

```{bash}
perl vcf2gl.pl filtered2x_frogs_uce.vcf
perl splitPops.pl filtered2x_frogs.gl
```
I then used `estpEM` (version 0.1) to estimate population allele frequencies (convergence tolerance = 0.001, 50 starts). This was done in the `PopGen` subdirectory.

```{perl}
#!/usr/bin/perl
#
# run estpEM
#

foreach $gl (@ARGV){
	$out = $gl;
	$out =~ s/\.gl// or die;
	system "estpEM -i $gl -o p_$out.txt -e 0.001 -m 50 -h 2\n";
}
```
We then computed pairwise Fst (as Nei's Gst, see[Nei 1973](https://www.pnas.org/doi/pdf/10.1073/pnas.70.12.3321)) for conspecific and heterospecific population pairs.
